### PR TITLE
Fix op-mainnet snapshot

### PR DESCRIPTION
**Pull request by deffrian**
Fixes Closes Resolves #
Pre-bedrock legacy transactions might not have signature

## Changes

- Adds op legacy tx decoder and signature verifier

#### What types of changes does your code introduce?

- [x] Bugfix (a non-breaking change that fixes an issue)
- [ ] New feature (a non-breaking change that adds functionality)
- [ ] Breaking change (a change that causes existing functionality not to work as expected)
- [ ] Optimization
- [ ] Refactoring
- [ ] Documentation update
- [ ] Build-related changes
- [ ] Other: _Description_

#### Requires testing

- [x] Yes
- [ ] No

#### If yes, did you write tests?

- [ ] Yes
- [x] No

#### Requires documentation update

- [ ] Yes
- [x] No

#### Requires explanation in Release Notes

- [ ] Yes
- [x] No

### CODE REVIEW

File: src/Nethermind/Nethermind.Optimism/OptimismLegacyTxDecoder.cs [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+        }
+        return base.DecodeSignature(v, rBytes, sBytes, fallbackSignature, rlpBehaviors);
+    }
+}
+
+public class OptimismLegacyTxValidator(OptimismSpecHelper specHelper) : ITxValidator
+{
+    public ValidationResult IsWellFormed(Transaction transaction, IReleaseSpec releaseSpec, BlockHeader? header = null)
+    {
+        if (header is null || specHelper.IsBedrock(header))
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Optimism/OptimismLegacyTxDecoder.cs:31]: Could IsBedrock be in `IReleaseSpec`, then we don't have to pass header?

`IReleaseSpec` was created to have all fork information.

**Review by LukaszRozmej — CHANGES_REQUESTED** [review]

File: src/Nethermind/Nethermind.Optimism/OptimismLegacyTxDecoder.cs (hunk 1 above)
Comment (deffrian) [reply src/Nethermind/Nethermind.Optimism/OptimismLegacyTxDecoder.cs:31]: we can check if eip1559 is enabled. Not sure it's good idea. But as I know ReleaseSpec is not aware of optimism forks

Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Optimism/OptimismLegacyTxDecoder.cs]: Would be good to make IReleaseSpec extendable by plugins then

**Review by LukaszRozmej — APPROVED** [review]

File: src/Nethermind/Nethermind.Optimism/OptimismLegacyTxDecoder.cs [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+}
+
+public class OptimismLegacyTxValidator : ITxValidator
+{
+    public ValidationResult IsWellFormed(Transaction transaction, IReleaseSpec releaseSpec)
+    {
+        // In op bedrock eip1559 activated with bedrock
+        if (releaseSpec.IsEip1559Enabled)
+        {
+            return transaction.Signature is null ? new ValidationResult("Empty signature") : ValidationResult.Success;
```
Comment (emlautarom1) [inline src/Nethermind/Nethermind.Optimism/OptimismLegacyTxDecoder.cs:34]: Before we were not performing any validations on Deposit transactions, but now we require that post-Bedrock transactions contain a non-null signature. Is this correct? If so, shouldn't we also validate the signature (using `SignatureTxValidator`)?

Comment (deffrian) [reply src/Nethermind/Nethermind.Optimism/OptimismLegacyTxDecoder.cs:34]: if signature is null here it means that while decoding signature in `DecodeSignature` we got inside  `if (v == 0 && rBytes.IsEmpty && sBytes.IsEmpty)`. Without this pr we would crash inside `DecodeSignature`.

Comment (emlautarom1) [reply src/Nethermind/Nethermind.Optimism/OptimismLegacyTxDecoder.cs:34]: That makes sense, what I'm missing is why we don't validate the signature of a Legacy transaction post bedrock. More so, why are we not performing the standard Legacy validations:

https://github.com/NethermindEth/nethermind/blob/d00410fbba8dd79d287c428c9aa4b02f399f9737/src/Nethermind/Nethermind.Consensus/Validators/TxValidator.cs#L24-L30

With these changes, Legacy transactions that are processed in Optimism are not validated at all, when previously they went through several checks.

Comment (deffrian) [reply src/Nethermind/Nethermind.Optimism/OptimismLegacyTxDecoder.cs:34]: Oh, that means I made a mistake. But why we don't check DepositTransaction?
`api.RegisterTxType<OptimismTransactionForRpc>(new OptimismTxDecoder<Transaction>(), Always.Valid);`

Comment (emlautarom1) [reply src/Nethermind/Nethermind.Optimism/OptimismLegacyTxDecoder.cs:34]: It dates back to the original Optimism plugin PR where the `OptimismTxValidator` performed no checks on Deposit transactions

https://github.com/NethermindEth/nethermind/blob/6edd397acfea5f262796db5a0f15592cf5d6cf53/src/Nethermind/Nethermind.Optimism/OptimismTxValidator.cs#L20-L26

We didn't change the logic in the Transaction refactoring, so the current validator for Deposit transactions always succeeds.

Comment (emlautarom1) [reply src/Nethermind/Nethermind.Optimism/OptimismLegacyTxDecoder.cs:34]: Addressed in https://github.com/NethermindEth/nethermind/pull/7887.

**Pull request merged**